File: java/shared/java/org/mochi/libmochi/zkgroup/backups/BackupAuthCredential.java
//
// Copyright 2023 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.mochi.libmochi.zkgroup.backups;

import static org.mochi.libmochi.internal.FilterExceptions.filterExceptions;
import static org.mochi.libmochi.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import org.mochi.libmochi.internal.Native;
import org.mochi.libmochi.zkgroup.GenericServerPublicParams;
import org.mochi.libmochi.zkgroup.InvalidInputException;
import org.mochi.libmochi.zkgroup.internal.ByteArray;

public final class BackupAuthCredential extends ByteArray {

  public BackupAuthCredential(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class,
        () -> Native.BackupAuthCredential_CheckValidContents(contents));
  }

  public BackupAuthCredentialPresentation present(GenericServerPublicParams serverParams) {
    return present(serverParams, new SecureRandom());
  }

  public BackupAuthCredentialPresentation present(
      GenericServerPublicParams serverParams, SecureRandom secureRandom) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    final byte[] newContents =
        filterExceptions(
            () ->
                Native.BackupAuthCredential_PresentDeterministic(
                    getInternalContentsForJNI(), serverParams.getInternalContentsForJNI(), random));

    try {
      return new BackupAuthCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public byte[] getBackupId() {
    return Native.BackupAuthCredential_GetBackupId(getInternalContentsForJNI());
  }

  public BackupLevel getBackupLevel() {
    return BackupLevel.fromValue(
        Native.BackupAuthCredential_GetBackupLevel(getInternalContentsForJNI()));
  }
}

File: swift/Sources/LibMochiClient/zkgroup/BackupAuthCredential.swift
//
// Copyright 2023 Mochi Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import MochiFfi

public class BackupAuthCredential: ByteArray {
    public required init(contents: [UInt8]) throws {
        try super.init(contents, checkValid: mochi_backup_auth_credential_check_valid_contents)
    }

    public func present(serverParams: GenericServerPublicParams) -> BackupAuthCredentialPresentation {
        return failOnError {
            self.present(serverParams: serverParams, randomness: try .generate())
        }
    }

    public func present(serverParams: GenericServerPublicParams, randomness: Randomness) -> BackupAuthCredentialPresentation {
        return failOnError {
            try withUnsafeBorrowedBuffer { contents in
                try serverParams.withUnsafeBorrowedBuffer { serverParams in
                    try randomness.withUnsafePointerToBytes { randomness in
                        try invokeFnReturningVariableLengthSerialized {
                            mochi_backup_auth_credential_present_deterministic($0, contents, serverParams, randomness)
                        }
                    }
                }
            }
        }
    }

    public var backupID: [UInt8] {
        return failOnError {
            try withUnsafeBorrowedBuffer { contents in
                try invokeFnReturningFixedLengthArray {
                    mochi_backup_auth_credential_get_backup_id($0, contents)
                }
            }
        }
    }

    public var backupLevel: BackupLevel {
        return failOnError {
            let rawValue = try withUnsafeBorrowedBuffer { contents in
                try invokeFnReturningInteger {
                    mochi_backup_auth_credential_get_backup_level($0, contents)
                }
            }
            guard let backupLevel = BackupLevel(rawValue: rawValue) else {
                throw MochiError.internalError("Invalid BackupLevel \(rawValue)")
            }
            return backupLevel
        }
    }
}
